static/_posts/2024-04-01-vgaterm-intro-001.md

```diff
@@ -0,0 +1,171 @@
+
+### VGA Timing
+
+VGA comes in several modes, and we target the 640x480, 60Hz frame rate mode - the base standard. VGA has 3 analog signals between 0 and 0.7 volts representing the amount of Red Green and Blue per pixel. There are also two timing signals `H_SYNC` and `V_SYNC` that tell the monitor what resolution mode to be in and when to sample the color signals to actually display a pixel. When the monitor is actively sampling colors and displaying pixels we say that this portion of the frame is `visible`. There is time during a complete frame when this isn't the case and this is the _blanking area_. Lastly, the _pixel clock_ speed is the rate at which pixels are shown in the visible portion, and in our case that's 25.175 MHz. In the diagram above, I've labeled a signal `CLK_VIS`, and this is the _pixel clock_ signal during the `visible` time. To see detailed VGA timing information, see [tiny vga](http://tinyvga.com/vga-timing/640x480@60Hz).
+
+:----:|:-----:
+![Timing Board front](/assets/images/vgaterm-timing-front.jpg) | ![Timing Board back](/assets/images/vgaterm-timing-back.jpg)
+<span style="color: {{ site.caption_color }};">Vgaterm Timing circuit top</span> | <span style="color: {{ site.caption_color }};">Vgaterm Timing cicuit bottom</span>
+
+VGA timing is controlled by the "timing board" circuit. This establishes the `H_SYNC` and `V_SYNC` signals that VGA requires, the `START` signal which alerts the CPU to begin  emitting pixels to the FIFO, and the `visible` flag, which is used to compute `CLK_VIS` to read pixel bytes out of the FIFO.
+
+### Pixels Out
+Every frame (60 times a second) the esp32c3 microcontroller needs to output 256,000 bytes (640x400 pixels, 1 byte per pixel) to hardware so they may be displayed. We decided that this is the most important software role as too much jitter in timing will cause artifacts and wobbling in the piture.
+
+To cross the clock domains, from the CPU SPI to VGA pixel clock, we used a FIFO memory chip. This lets us write bytes from the CPU while VGA hardware reads bytes exactly when it needs them.
+
+{% include figure.html image="/assets/images/fifo-chip-highlight.jpg" width=450 position="right" caption="IDT 7205 FIFO chip" %}
+
+This worked well, but there are a few constraints we had to consider when designing a circuit around the FIFO:
+* The FIFO read and write setup and hold times had to be obeyed by `CLK_VIS` and the SPI clock signals respectively
+* The SPI only supported clock speeds in power-of-two divisions of 80 MHz (80, 40, 20, 10, etc)
+* If we emitted pixels too quickly, and VGA hardware could not read from the FIFO fast enough, we would blow the queue and drop pixels if the CPU didn't pause writing periodically.
+* If we emitted pixels too slowly, and VGA hardware consumed pixels faster than was being written, then we'd have to start loading pixels before VGA started consuming them.
```

Comment:
```suggestion
* If we emitted pixels too quickly, and VGA hardware did not read from the FIFO fast enough, we would blow the queue and drop pixels
* If we emitted pixels too slowly, and VGA hardware consumed pixels faster than was being written, then we would also be missing pixels
```

Pausing periodically was an effect, IIRC, of the DMA sizing and not a design goal: but loading pixels before the VGA started _was_ a design goal.

It's good to give the reader a sense of the difficulty in lining up the average read rates, but I don't think you need to go into the "20MHz, except when it wasn't" and "21.175MHz, but a whole bunch of the pixels are timing controlled and have to be black" details here.